rfc/plugins.md

```diff
@@ -0,0 +1,300 @@
+
+Tightly-coupled service and plugin releases is a poor developer and operator story, but an acceptable shortcoming for an initial release. By using Front50 as the source of truth, we’ll be fairly well equipped to make an acceptable runtime plugin loading story.
+
+Assuming a world where in-process plugins can be upgraded at runtime, services will need a background process to download these new plugins whenever a delta is detected in Front50, and the plugin refreshed. PF4J offers an abstraction for this reloading mechanism, but we would need to figure out how Spring integrations are handled. For example, it’s possible we may need to make the constraint that exported beans from a plugin are not refreshed on plugin up/down grade at least initially.
+
+#### Remote
+Remote plugin upgrade process is not much different than iterating a remote service, since that’s what it is.
+
+When a remote plugin is released, its internal functionality will change immediately, just as it would be expected for a remote service. When the plugin starts, it must send a registration event to Spinnaker. This registration event would contain, at least:
+
+* Plugin ID
+* Other plugin metadata (author, description, version, etc).
+* What extension points it integrates with
+
+Upon receiving this registration event, Spinnaker would then update the Front50 registry with the new information and downstream services would re-bootstrap the plugin, adding or removing it from extension points as necessary.
+
+### Deployment of Remote Plugins
+
+Halyard based Spinnaker installations will use halyard to deploy remote plugins, similar to other spinnaker services.
+Halyard will also be responsible for registering remote plugins with Front50.
+
+For non-halyard based Spinnaker installations, its up to the operator to deploy and register remote plugins.
+
+### Deployment of In-Process Plugins
+
+Downloading plugin resources is necessary for in-process plugins.
+Services need to have plugin resources available locally in order to load them into their process.
+Plugin resources are hosted on a remote server and must be downloaded in order to use them.
+The existing method of downloading plugin resources involves the service itself downloading plugins via a given URI.
+While this works for the java services, it is not a general solution.
+Furthermore, downloading of plugin resources is a deployment concern and should not be part of the runtime environment.
+
+#### Kubernetes Environment
+
+Halyard can be used for deploying plugins to Kubernetes environments.
+To deploy plugins, Halyard will follow the same pattern as kubernetes regarding downloading resources prior to runtime.
+Downloading of resources will be done through init containers and volume mounts that are then shared with their respective service.
+By using init containers to download resources, we can standardize on one method of deploying plugins to spinnaker services.
```

Comment:
Hmm... I can see the value in this, but if we're going to support runtime downloads as discussed above, there will need to be some logic in the core service(s) to do that, so we'll end up with two different ways to download and install plugins, no? Do we plan to start without runtime download support and add it later?

Reply:
Yes, we'll have two different ways to get plugins in a K8s environment. Unfortunately, this is pretty well out of our control at this point (@link108 keep me honest?). Runtime download support will not be initially supported (because we need to build it), but it will be an early feature as it's critical to how Deck's plugin model will work.

Reply:
I wouldn't say its set in stone, I'd prefer to have one method of deploying plugins. I'm all for going the remote downloadable route, with plugin metadata stored in Front50 👍 After some of the PR updates, I can more easily grok the remote-downloadable solution and can update this accordingly.

I do think Halyard still has a part to play in this, I view plugins as part of a spinnaker runtime and as such, should be part of the hal configs. Rather than deploy the plugins, Halyard would update Front50 with the plugin metadata, provided in the manifest.yml file.